gemhardware/include/gem/hw/GEMHwDevice.h

```diff
@@ -581,10 +525,45 @@ namespace gem {
        * @brief Performs basic setup for the device
        * sets connection details (OBSOLETE)
        * sets logging level to Error
-       * Not inherited
+       * Not inherited, but calls a pure virtual function...
        **/
       void setup(std::string const& deviceName);
 
+      /**
+       * @brief Performs a check on the RPC response to verify whether there is an `error` key set
+       *        Every method that makes an RPC call *must* use this function to check the response
+       *
+       * @param caller should be the name of the function, only used in the error message
+       **/
+      void checkRPCResponse(std::string const& caller) const {
+        if (rsp.get_key_exists("error")) {
+          std::stringstream errmsg;
+          errmsg << rsp.get_string("error");
+          CMSGEMOS_ERROR(caller << ": " << errmsg.str());
+          XCEPT_RAISE(gem::hw::exception::RPCMethodError, errmsg.str());
+        }
+      }
+
+      /**
+       * @brief Extracts the device parameters from the device name
+       *
+       * @param deviceName should be of the format: gem-shelfXX-amcYY
+       *        optionally including -optohybridZZ
+       * @param index selects which of XX, YY, ZZ to return
+       **/
+      static uint8_t extractDeviceID(std::string const& deviceName, uint8_t const& index) {
```

Comment:
* Not a good candidate for an `inline` function (too complex), move implementation to `.cc`.
* Finding the `shelf`, `amc` and `optohybrid` parameters would require parsing the name three times. Why not return `std::vector<uint8_t>`?
* Looks like the implementation also supports an extra `-vfatVV`, why not state it in the docs?

Reply:
> * Finding the `shelf`, `amc` and `optohybrid` parameters would require parsing the name three times. Why not return `std::vector<uint8_t>`?

It's a possibility... how I use it didn't propose that option.  If the use case exists, probably better to implement your idea of a `GEMHwAddress` `struct`.

> * Looks like the implementation also supports an extra `-vfatVV`, why not state it in the docs?

I am not envisioning supporting a `HwVFAT` type class (which was never really done in the core code anyway), though this is open for *some* discussion.
The reasoning is that minimizing the number of RPC connections (and also `uhal` devices) is in our best interests, and since what we really have are `BackEndDevice` (`AMC`) and `FrontEndDevice` (covering `OptoHybrid`, `GBTx`, VFAT`, `SCA`).
Current usage in `cmsgemos` managers, tried at one point to have `HwVFAT` devices, but it never really made good sense, since most of what we wanted to do was broadcast type anyway, and now configurations will be streamed per front-end link anyway.

This is something that will have to be addressed when certain python code is migrated, as @bdorney created `HwVFAT` devices for some tasks, but I don't see a problem migrating this.
In the worst case, we can indeed create `HwVFAT` devices, but I would only do it as a return type from a `HwOptoHybrid` function, *only* as a create-use-destroy (resource) model, and probably *only* as an `XHALInterface` (i.e., RPC only, to avoid the address table parsing cost).

Reply:
>> Finding the shelf, amc and optohybrid parameters would require parsing the name three times. Why not return std::vector<uint8_t>?
>
> It's a possibility... how I use it didn't propose that option. If the use case exists, probably better to implement your idea of a GEMHwAddress struct.

If Umesh decides to use the naming system in the DB, I'll need this function (and the reverse, ie creating the name) for `gemonlinedb`.